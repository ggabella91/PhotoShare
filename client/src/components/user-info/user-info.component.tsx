import React, { useState } from 'react';
import { useHistory, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { createStructuredSelector } from 'reselect';

import { AppState } from '../../redux/root-reducer';

import { DeleteReactionReq } from '../../redux/post/post.types';

import {
  setCommentToDelete,
  setShowCommentOptionsModal,
  setShowPostEditForm,
} from '../../redux/post/post.actions';

import './user-info.styles.scss';

export enum StyleType {
  suggestion = 'suggestion',
  modal = 'modal',
  feed = 'feed',
  comment = 'comment',
}

export interface UserInfoAndOtherData {
  profilePhotoFileString: string;
  username: string;
  name: string;
  location: string;
  comment: string;
  commentDate?: Date | string;
  reactionId?: string;
  reactingUserId?: string;
}

interface UserInfoProps {
  styleType: StyleType;
  userInfoArray: UserInfoAndOtherData[];
  isCaption?: boolean;
  isCaptionOwner?: boolean;
  setCommentToDelete: typeof setCommentToDelete;
  setShowCommentOptionsModal: typeof setShowCommentOptionsModal;
  setShowPostEditForm: typeof setShowPostEditForm;
}

export const UserInfo: React.FC<UserInfoProps> = ({
  userInfoArray,
  styleType,
  isCaption,
  isCaptionOwner,
  setCommentToDelete,
  setShowCommentOptionsModal,
  setShowPostEditForm,
}) => {
  const [showCommentOptionsButtonForIdx, setShowCommentOptionsButtonForIdx] =
    useState({ show: false, idx: -1 });

  let history = useHistory();

  const handleSetCommentToDelete = (idx: number) => {
    const commentToDelete = userInfoArray[idx];
    if (commentToDelete.reactionId && commentToDelete.reactingUserId) {
      setCommentToDelete({
        reactionId: commentToDelete.reactionId,
        isLikeRemoval: false,
        reactingUserId: commentToDelete.reactingUserId,
      });

      setShowCommentOptionsModal(true);
    }
  };

  const handleClickCaptionOptions = () => {
    if (isCaptionOwner) {
      setShowPostEditForm(true);
    }
  };

  const userInfo = userInfoArray.map(
    (el: UserInfoAndOtherData, idx: number) => (
      <div
        className={`user-${styleType}-element`}
        key={idx}
        onClick={
          styleType === StyleType.suggestion
            ? () => {
                history.push(`/${el.username}`);
              }
            : () => {}
        }
        onMouseEnter={() =>
          setShowCommentOptionsButtonForIdx({ show: true, idx })
        }
        onMouseLeave={() =>
          setShowCommentOptionsButtonForIdx({ show: false, idx })
        }
      >
        <div className={`${styleType}-avatar`}>
          {el.profilePhotoFileString ? (
            <img
              className={`${styleType}-profile-photo`}
              src={`data:image/jpeg;base64,${el.profilePhotoFileString}`}
              alt='profile-pic'
            />
          ) : null}
          {!el.profilePhotoFileString ? (
            <div className={`${styleType}-photo-placeholder`}>
              <span className={`${styleType}-photo-placeholder-text`}>
                No photo
              </span>
            </div>
          ) : null}
        </div>
        <div className='user-data-and-date'>
          <div className={`${styleType}-username-and-other-data`}>
            {styleType !== StyleType.suggestion ? (
              <NavLink
                to={`/${el.username}`}
                className={`${styleType}-username`}
              >
                {el.username}
              </NavLink>
            ) : (
              <span className={`${styleType}-username`}>{el.username}</span>
            )}
            <span className={`${styleType}-name`}>{el.name}</span>
            <span className={`${styleType}-location`}>{el.location}</span>
            <span>{el.comment ? el.comment : null}</span>
          </div>
          {el.commentDate ? (
            <span className={`${styleType}-date`}>
              {new Date(el.commentDate).toDateString()}
            </span>
          ) : null}
        </div>
        <div
          className={`${
            showCommentOptionsButtonForIdx.show &&
            showCommentOptionsButtonForIdx.idx === idx
              ? ''
              : 'hide'
          } comment-options`}
        >
          {styleType === StyleType.comment && !isCaption ? (
            <span
              className='comment-ellipsis-button'
              onClick={() => handleSetCommentToDelete(idx)}
            >
              ...
            </span>
          ) : null}
          {styleType === StyleType.comment && isCaption ? (
            <span
              className='comment-ellipsis-button'
              onClick={handleClickCaptionOptions}
            >
              ...
            </span>
          ) : null}
        </div>
      </div>
    )
  );

  return <div className={`user-${styleType}-container`}>{userInfo}</div>;
};

interface LinkStateProps {}

const mapStateToProps = createStructuredSelector<AppState, LinkStateProps>({});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  setCommentToDelete: (deleteReactionReq: DeleteReactionReq) =>
    dispatch(setCommentToDelete(deleteReactionReq)),
  setShowCommentOptionsModal: (showCommentOptionsModal: boolean) =>
    dispatch(setShowCommentOptionsModal(showCommentOptionsModal)),
  setShowPostEditForm: (showPostEditForm: boolean) =>
    dispatch(setShowPostEditForm(showPostEditForm)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UserInfo);
