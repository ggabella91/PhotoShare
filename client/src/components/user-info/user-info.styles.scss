.user-suggestion-container {
  display: flex;
  flex-direction: column;
  position: absolute;
  width: 220px;
  max-height: 100px;
  background-color: white;
  overflow-y: scroll;
  overflow-x: hidden;
  border-radius: 4px;
  transform: translate(-10px);
  z-index: 20;

  .user-suggestion-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 220px;
    height: 50px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);

    &:hover {
      cursor: pointer;
    }

    .suggestion-avatar {
      position: relative;
      width: 50px;
      height: 50px;
      border-radius: 50%;

      .suggestion-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 32px;
        height: 32px;
        border-radius: 50%;
      }

      .suggestion-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 32px;
        height: 32px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;

        .suggestion-photo-placeholder-text {
          position: relative;
          font-size: 8px;
          color: black;
          margin-top: 50%;
          transform: translateY(-55%);
        }
      }
    }

    .suggestion-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 4px;
      width: 140px;
      height: 50px;
      justify-self: flex-end;
      text-align: left;

      .suggestion-username {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 420;
        color: black;
      }

      .suggestion-name {
        display: flex;
        position: relative;
        font-size: 12px;
        font-weight: 380;
        color: black;
      }

      .suggestion-location {
        display: none;
      }
    }
  }
}

.user-modal-container {
  display: flex;
  flex-direction: column;
  position: relative;
  width: inherit;
  max-height: 338px;
  background-color: white;
  border-radius: 4px;

  .user-modal-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 100%;
    height: 70px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);
    padding-top: 5px;
    padding-bottom: 5px;

    &:hover {
      cursor: pointer;
    }

    .modal-avatar {
      position: relative;
      width: 60px;
      height: 60px;
      border-radius: 50%;
      margin-left: 15px;
      margin-right: 2vw;

      .modal-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 45px;
        height: 45px;
        border-radius: 50%;
      }

      .modal-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 45px;
        height: 45px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .modal-photo-placeholder-text {
          position: relative;
          font-size: 10px;
          color: black;
          margin-top: 50%;
          margin-left: 50%;
          transform: translate(-50%, -50%);
        }
      }
    }

    .modal-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 6px;
      width: 75%;
      height: 60px;
      justify-self: flex-end;
      text-align: left;

      .modal-username {
        display: flex;
        position: relative;
        font-size: 16px;
        font-weight: 420;
        color: black;
      }

      .modal-name {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 380;
        color: black;
      }

      .modal-location {
        display: none;
      }
    }
  }
}

.user-feed-container {
  display: flex;
  flex-direction: column;
  position: relative;
  width: inherit;
  max-height: 338px;
  background-color: white;
  border-radius: 4px;

  .user-feed-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 100%;
    height: 60px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);
    padding-top: 5px;
    padding-bottom: 5px;

    &:hover {
      cursor: pointer;
    }

    .feed-avatar {
      position: relative;
      width: 50px;
      height: 50px;
      border-radius: 50%;
      margin-left: 15px;
      margin-right: 15px;
      padding-top: 5px;

      .feed-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
      }

      .feed-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .feed-photo-placeholder-text {
          position: relative;
          font-size: 10px;
          color: black;
          margin-top: 50%;
          margin-left: 50%;
          transform: translate(-50%, -50%);
        }
      }
    }

    .feed-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 6px;
      width: 75%;
      height: 60px;
      justify-self: flex-end;
      text-align: left;

      .feed-username {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 540;
        color: black;
      }

      .feed-name {
        display: none;
      }

      .feed-location {
        display: flex;
        position: relative;
        font-size: 12px;
        font-weight: 380;
        color: black;
      }
    }
  }
}

.user-comment-container {
  display: flex;
  flex-direction: column;
  position: relative;
  width: inherit;
  max-height: 338px;
  background-color: white;
  border-radius: 4px;

  .user-comment-element {
    display: flex;
    position: relative;
    justify-content: start;
    width: 100%;
    height: 60px;
    background-color: white;
    border-bottom: solid 1px rgb(180, 180, 180);
    padding-top: 5px;
    padding-bottom: 5px;

    &:hover {
      cursor: pointer;
    }

    .comment-avatar {
      position: relative;
      width: 50px;
      height: 50px;
      border-radius: 50%;
      margin-left: 15px;
      margin-right: 15px;
      padding-top: 5px;

      .comment-profile-photo {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
      }

      .comment-photo-placeholder {
        position: relative;
        margin-left: 50%;
        margin-top: 50%;
        transform: translate(-50%, -50%);
        width: 40px;
        height: 40px;
        border-radius: 50%;
        border: solid 1px;
        border-color: black;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;

        .comment-photo-placeholder-text {
          position: relative;
          font-size: 10px;
          color: black;
          margin-top: 50%;
          margin-left: 50%;
          transform: translate(-50%, -50%);
        }
      }
    }

    .comment-username-and-other-data {
      display: flex;
      position: relative;
      flex-direction: column;
      padding-top: 6px;
      width: 75%;
      height: 60px;
      justify-self: flex-end;
      text-align: left;

      .comment-username {
        display: flex;
        position: relative;
        font-size: 14px;
        font-weight: 540;
        color: black;
      }

      .comment-name {
        display: none;
      }

      .comment-location {
        display: none;
      }

      .comment-text {
        display: flex;
        position: relative;
        font-size: 12;
        font-weight: 400;
      }
    }

    .comment-date {
      display: flex;
      position: relative;
      font-size: 10;
      font-weight: 380;
      margin-top: 5px;
    }
  }
}
