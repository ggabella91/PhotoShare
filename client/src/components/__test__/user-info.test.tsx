import { shallow } from 'enzyme';
import React from 'react';

import { UserInfo } from '../user-info/user-info.component';

it('renders a user-suggestions component', () => {
  const userInfoWrapper = shallow(<UserInfo userInfoArray={[]} />);

  expect(userInfoWrapper).toMatchSnapshot();
});
