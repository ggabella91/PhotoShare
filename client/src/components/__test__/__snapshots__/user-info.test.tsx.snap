// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a user-suggestions component 1`] = `
<div
  className="user-suggestions-dropdown"
/>
`;
